packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+  scope?: string
+  maxRetries?: number
+  httpClient?: HTTPClient
+}
+
+export class TokenManager {
+  private alg = 'RS256' as const
+  private grantType = 'client_credentials' as const
+  private clientAssertionType =
+    'urn:ietf:params:oauth:client-assertion-type:jwt-bearer' as const
+  private clientId: string
+  private clientKey: Buffer
+  private keyId: string
+  private scope: string
+  private authServer: string
+  private httpClient: HTTPClient
+  private maxRetries: number
+  private clockSkewInSeconds = 0
+
+  private controller: AbortController
+  private signal: AbortSignal
+
+  private accessToken?: AccessToken
+  private tokenEmitter = new Emitter<{
+    access_token: [{ token: AccessToken } | { error: unknown }]
+  }>()
+  private retryCount: number
+  private lastError: any
+  private pollerTimer?: ReturnType<typeof setTimeout>
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
+    this.maxRetries = props.maxRetries ?? 3
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.controller = new AbortController()
+    this.signal = this.controller.signal
+    this.signal.addEventListener('abort', () => {
+      clearTimeout(this.pollerTimer)
+    })
+    this.retryCount = 0
+  }
+
+  async pollerLoop() {
+    let timeUntilRefreshInMs = 25
+    let response: HTTPResponse
+
+    try {
+      response = await this.requestAccessToken()
+    } catch (err) {
+      // Error without a status code - likely networking, retry
+      this.retryCount++
+      this.lastError = err
+
+      if (this.retryCount % this.maxRetries == 0) {
+        this.tokenEmitter.emit('access_token', { error: this.lastError })
+      }
+
+      this.pollerTimer = setTimeout(
+        this.pollerLoop.bind(this),
+        backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      ).unref()
+      return
+    }
+
+    if (response.headers !== undefined && response.headers.Date != undefined) {
+      const serverTime = Date.parse(response.headers.Date)
+      const skew = Date.now() - serverTime
+      if (this.clockSkewInSeconds == 0) {
+        this.clockSkewInSeconds = skew
+      } else {
+        this.clockSkewInSeconds = (this.clockSkewInSeconds + skew) / 2
+      }
+    }
+
+    // Handle status codes!
+    if (response.status === 200) {
+      let body: any
+      if (typeof response.text != 'function') {
+        this.tokenEmitter.emit('access_token', {
+          error: new Error(
+            'HTTPClient does not implement response.text method'
+          ),
+        })
+        clearTimeout(this.pollerTimer)
```

Comment:
Did you find this was needed, or just being cautious? I would have expected this to not be necessary as long as we return before calling setTimeout again.

Reply:
cautious/paranoid.  I think I wrote this early on before it had solidified, I don't think it's necessary.